File: pkg/codegen/testing/test/testdata/aws-fargate-pp/go/aws-fargate.go
package main

import (
	"encoding/json"

	"github.com/pulumi/pulumi-aws/sdk/v5/go/aws/ec2"
	"github.com/pulumi/pulumi-aws/sdk/v5/go/aws/ecs"
	"github.com/pulumi/pulumi-aws/sdk/v5/go/aws/elasticloadbalancingv2"
	"github.com/pulumi/pulumi-aws/sdk/v5/go/aws/iam"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		vpc, err := ec2.LookupVpc(ctx, &ec2.LookupVpcArgs{
			Default: pulumi.BoolRef(true),
		}, nil)
		if err != nil {
			return err
		}
		subnets, err := ec2.GetSubnetIds(ctx, &ec2.GetSubnetIdsArgs{
			VpcId: vpc.Id,
		}, nil)
		if err != nil {
			return err
		}
		webSecurityGroup, err := ec2.NewSecurityGroup(ctx, "webSecurityGroup", &ec2.SecurityGroupArgs{
			VpcId: pulumi.String(vpc.Id),
			Egress: []ec2.SecurityGroupEgressArgs{
				{
					Protocol: pulumi.String("-1"),
					FromPort: pulumi.Int(0),
					ToPort:   pulumi.Int(0),
					CidrBlocks: []pulumi.String{
						pulumi.String("0.0.0.0/0"),
					},
				},
			},
			Ingress: []ec2.SecurityGroupIngressArgs{
				{
					Protocol: pulumi.String("tcp"),
					FromPort: pulumi.Int(80),
					ToPort:   pulumi.Int(80),
					CidrBlocks: []pulumi.String{
						pulumi.String("0.0.0.0/0"),
					},
				},
			},
		})
		if err != nil {
			return err
		}
		cluster, err := ecs.NewCluster(ctx, "cluster", nil)
		if err != nil {
			return err
		}
		tmpJSON0, err := json.Marshal(map[string]interface{}{
			"Version": "2008-10-17",
			"Statement": []map[string]interface{}{
				map[string]interface{}{
					"Sid":    "",
					"Effect": "Allow",
					"Principal": map[string]interface{}{
						"Service": "ecs-tasks.amazonaws.com",
					},
					"Action": "sts:AssumeRole",
				},
			},
		})
		if err != nil {
			return err
		}
		json0 := string(tmpJSON0)
		taskExecRole, err := iam.NewRole(ctx, "taskExecRole", &iam.RoleArgs{
			AssumeRolePolicy: pulumi.String(json0),
		})
		if err != nil {
			return err
		}
		_, err = iam.NewRolePolicyAttachment(ctx, "taskExecRolePolicyAttachment", &iam.RolePolicyAttachmentArgs{
			Role:      taskExecRole.Name,
			PolicyArn: pulumi.String("arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy"),
		})
		if err != nil {
			return err
		}
		webLoadBalancer, err := elasticloadbalancingv2.NewLoadBalancer(ctx, "webLoadBalancer", &elasticloadbalancingv2.LoadBalancerArgs{
			Subnets: toPulumiStringArray(subnets.Ids),
			SecurityGroups: []pulumi.String{
				webSecurityGroup.ID(),
			},
		})
		if err != nil {
			return err
		}
		webTargetGroup, err := elasticloadbalancingv2.NewTargetGroup(ctx, "webTargetGroup", &elasticloadbalancingv2.TargetGroupArgs{
			Port:       80,
			Protocol:   "HTTP",
			TargetType: "ip",
			VpcId:      pulumi.String(vpc.Id),
		})
		if err != nil {
			return err
		}
		webListener, err := elasticloadbalancingv2.NewListener(ctx, "webListener", &elasticloadbalancingv2.ListenerArgs{
			LoadBalancerArn: webLoadBalancer.Arn,
			Port:            80,
			DefaultActions: elasticloadbalancingv2.ListenerDefaultActionArray{
				&elasticloadbalancingv2.ListenerDefaultActionArgs{
					Type:           pulumi.String("forward"),
					TargetGroupArn: webTargetGroup.Arn,
				},
			},
		})
		if err != nil {
			return err
		}
		tmpJSON1, err := json.Marshal([]map[string]interface{}{
			map[string]interface{}{
				"name":  "my-app",
				"image": "nginx",
				"portMappings": []map[string]interface{}{
					map[string]interface{}{
						"containerPort": 80,
						"hostPort":      80,
						"protocol":      "tcp",
					},
				},
			},
		})
		if err != nil {
			return err
		}
		json1 := string(tmpJSON1)
		appTask, err := ecs.NewTaskDefinition(ctx, "appTask", &ecs.TaskDefinitionArgs{
			Family:      pulumi.String("fargate-task-definition"),
			Cpu:         "256",
			Memory:      "512",
			NetworkMode: "awsvpc",
			RequiresCompatibilities: []pulumi.String{
				pulumi.String("FARGATE"),
			},
			ExecutionRoleArn:     taskExecRole.Arn,
			ContainerDefinitions: pulumi.String(json1),
		})
		if err != nil {
			return err
		}
		_, err = ecs.NewService(ctx, "appService", &ecs.ServiceArgs{
			Cluster:        cluster.Arn,
			DesiredCount:   5,
			LaunchType:     "FARGATE",
			TaskDefinition: appTask.Arn,
			NetworkConfiguration: &*ecs.ServiceNetworkConfigurationArgs{
				AssignPublicIp: true,
				Subnets:        toPulumiStringArray(subnets.Ids),
				SecurityGroups: []pulumi.String{
					webSecurityGroup.ID(),
				},
			},
			LoadBalancers: []ecs.ServiceLoadBalancerArgs{
				{
					TargetGroupArn: webTargetGroup.Arn,
					ContainerName:  pulumi.String("my-app"),
					ContainerPort:  pulumi.Int(80),
				},
			},
		}, pulumi.DependsOn([]pulumi.Resource{
			webListener,
		}))
		if err != nil {
			return err
		}
		ctx.Export("url", webLoadBalancer.DnsName)
		return nil
	})
}
func toPulumiStringArray(arr []string) pulumi.StringArray {
	var pulumiArr pulumi.StringArray
	for _, v := range arr {
		pulumiArr = append(pulumiArr, pulumi.String(v))
	}
	return pulumiArr
}

File: pkg/codegen/dotnet/testdata/aws-fargate-pp/aws-fargate.cs
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using Pulumi;
using Aws = Pulumi.Aws;

return await Deployment.RunAsync(() => 
{
    // Read the default VPC and public subnets, which we will use.
    var vpc = Aws.Ec2.GetVpc.Invoke(new()
    {
        Default = true,
    });

    var subnets = Aws.Ec2.GetSubnetIds.Invoke(new()
    {
        VpcId = vpc.Apply(getVpcResult => getVpcResult.Id),
    });

    // Create a security group that permits HTTP ingress and unrestricted egress.
    var webSecurityGroup = new Aws.Ec2.SecurityGroup("webSecurityGroup", new()
    {
        VpcId = vpc.Apply(getVpcResult => getVpcResult.Id),
        Egress = new[]
        {
            new Aws.Ec2.Inputs.SecurityGroupEgressArgs
            {
                Protocol = "-1",
                FromPort = 0,
                ToPort = 0,
                CidrBlocks = new[]
                {
                    "0.0.0.0/0",
                },
            },
        },
        Ingress = new[]
        {
            new Aws.Ec2.Inputs.SecurityGroupIngressArgs
            {
                Protocol = "tcp",
                FromPort = 80,
                ToPort = 80,
                CidrBlocks = new[]
                {
                    "0.0.0.0/0",
                },
            },
        },
    });

    // Create an ECS cluster to run a container-based service.
    var cluster = new Aws.Ecs.Cluster("cluster");

    // Create an IAM role that can be used by our service's task.
    var taskExecRole = new Aws.Iam.Role("taskExecRole", new()
    {
        AssumeRolePolicy = JsonSerializer.Serialize(new Dictionary<string, object?>
        {
            ["Version"] = "2008-10-17",
            ["Statement"] = new[]
            {
                new Dictionary<string, object?>
                {
                    ["Sid"] = "",
                    ["Effect"] = "Allow",
                    ["Principal"] = new Dictionary<string, object?>
                    {
                        ["Service"] = "ecs-tasks.amazonaws.com",
                    },
                    ["Action"] = "sts:AssumeRole",
                },
            },
        }),
    });

    var taskExecRolePolicyAttachment = new Aws.Iam.RolePolicyAttachment("taskExecRolePolicyAttachment", new()
    {
        Role = taskExecRole.Name,
        PolicyArn = "arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy",
    });

    // Create a load balancer to listen for HTTP traffic on port 80.
    var webLoadBalancer = new Aws.ElasticLoadBalancingV2.LoadBalancer("webLoadBalancer", new()
    {
        Subnets = subnets.Apply(getSubnetIdsResult => getSubnetIdsResult.Ids),
        SecurityGroups = new[]
        {
            webSecurityGroup.Id,
        },
    });

    var webTargetGroup = new Aws.ElasticLoadBalancingV2.TargetGroup("webTargetGroup", new()
    {
        Port = 80,
        Protocol = "HTTP",
        TargetType = "ip",
        VpcId = vpc.Apply(getVpcResult => getVpcResult.Id),
    });

    var webListener = new Aws.ElasticLoadBalancingV2.Listener("webListener", new()
    {
        LoadBalancerArn = webLoadBalancer.Arn,
        Port = 80,
        DefaultActions = new[]
        {
            new Aws.ElasticLoadBalancingV2.Inputs.ListenerDefaultActionArgs
            {
                Type = "forward",
                TargetGroupArn = webTargetGroup.Arn,
            },
        },
    });

    // Spin up a load balanced service running NGINX
    var appTask = new Aws.Ecs.TaskDefinition("appTask", new()
    {
        Family = "fargate-task-definition",
        Cpu = "256",
        Memory = "512",
        NetworkMode = "awsvpc",
        RequiresCompatibilities = new[]
        {
            "FARGATE",
        },
        ExecutionRoleArn = taskExecRole.Arn,
        ContainerDefinitions = JsonSerializer.Serialize(new[]
        {
            new Dictionary<string, object?>
            {
                ["name"] = "my-app",
                ["image"] = "nginx",
                ["portMappings"] = new[]
                {
                    new Dictionary<string, object?>
                    {
                        ["containerPort"] = 80,
                        ["hostPort"] = 80,
                        ["protocol"] = "tcp",
                    },
                },
            },
        }),
    });

    var appService = new Aws.Ecs.Service("appService", new()
    {
        Cluster = cluster.Arn,
        DesiredCount = 5,
        LaunchType = "FARGATE",
        TaskDefinition = appTask.Arn,
        NetworkConfiguration = new Aws.Ecs.Inputs.ServiceNetworkConfigurationArgs
        {
            AssignPublicIp = true,
            Subnets = subnets.Apply(getSubnetIdsResult => getSubnetIdsResult.Ids),
            SecurityGroups = new[]
            {
                webSecurityGroup.Id,
            },
        },
        LoadBalancers = new[]
        {
            new Aws.Ecs.Inputs.ServiceLoadBalancerArgs
            {
                TargetGroupArn = webTargetGroup.Arn,
                ContainerName = "my-app",
                ContainerPort = 80,
            },
        },
    }, new CustomResourceOptions
    {
        DependsOn =
        {
            webListener,
        },
    });

    return new Dictionary<string, object?>
    {
        ["url"] = webLoadBalancer.DnsName,
    };
});

